raftexample restart without leader

single raft cluster. First start can vote a leader, but restart cannot vote leader. It always follower

zouxu:raftexample zouxu$ ./raftexample --id 1 --cluster http://127.0.0.1:12379 --port 12380
raft2016/02/10 15:45:17 INFO: 1 became follower at term 0
raft2016/02/10 15:45:17 INFO: newRaft 1 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2016/02/10 15:45:17 INFO: 1 became follower at term 1
raft2016/02/10 15:45:18 INFO: 1 is starting a new election at term 1
raft2016/02/10 15:45:18 INFO: 1 became candidate at term 2
raft2016/02/10 15:45:18 INFO: 1 received vote from 1 at term 2
raft2016/02/10 15:45:18 INFO: 1 became leader at term 2
raft2016/02/10 15:45:18 INFO: raft.node: 1 elected leader 1 at term 2
killall raftexample
^Czouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ 
zouxu:raftexample zouxu$ ./raftexample --id 1 --cluster http://127.0.0.1:12379 --port 12380
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x28 pc=0x5144]

goroutine 14 [running]:
main.(*raftNode).publishEntries(0xc820092c80, 0xc8200c4240, 0x3, 0x4, 0x0)
	/Users/zouxu/.golang/gopath/src/fish.red/zouxu/goall/other/raftexample/raft.go:107 +0x354
main.(*raftNode).replayWAL(0xc820092c80, 0xd)
	/Users/zouxu/.golang/gopath/src/fish.red/zouxu/goall/other/raftexample/raft.go:157 +0x1dc
main.(*raftNode).startRaft(0xc820092c80)
	/Users/zouxu/.golang/gopath/src/fish.red/zouxu/goall/other/raftexample/raft.go:173 +0x67
created by main.newRaftNode
	/Users/zouxu/.golang/gopath/src/fish.red/zouxu/goall/other/raftexample/raft.go:82 +0x450

goroutine 1 [chan receive]:
main.(*kvstore).readCommits(0xc8200c2fc0, 0xc820018a20, 0xc820018a80)
	/Users/zouxu/.golang/gopath/src/fish.red/zouxu/goall/other/raftexample/kvstore.go:61 +0x6e
main.newKVStore(0xc820018960, 0xc820018a20, 0xc820018a80, 0xc820063dc8)
	/Users/zouxu/.golang/gopath/src/fish.red/zouxu/goall/other/raftexample/kvstore.go:39 +0xd4
main.serveHttpKVAPI(0x305c, 0xc820018960, 0xc8200189c0, 0xc820018a20, 0xc820018a80)
	/Users/zouxu/.golang/gopath/src/fish.red/zouxu/goall/other/raftexample/httpapi.go:110 +0x92
main.main()
	/Users/zouxu/.golang/gopath/src/fish.red/zouxu/goall/other/raftexample/main.go:40 +0x28f

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 9 [chan receive]:
github.com/coreos/etcd/pkg/logutil.(*MergeLogger).outputLoop(0xc8200ccbc0)
	/Users/zouxu/.golang/gopath/src/github.com/coreos/etcd/pkg/logutil/merge_logger.go:174 +0x7c
created by github.com/coreos/etcd/pkg/logutil.NewMergeLogger
	/Users/zouxu/.golang/gopath/src/git

--- a/contrib/raftexample/httpapi.go
+++ b/contrib/raftexample/httpapi.go
@@ -18,6 +18,7 @@ import (
 	"io/ioutil"
 	"log"
 	"net/http"
+	"os"
 	"strconv"
 
 	"github.com/coreos/etcd/raft/raftpb"
@@ -103,6 +104,14 @@ func (h *httpKVAPI) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 // serveHttpKVAPI starts a key-value server with a GET/PUT API and listens.
 func serveHttpKVAPI(port int, proposeC chan<- string, confChangeC chan<- raftpb.ConfChange,
 	commitC <-chan *string, errorC <-chan error) {
+
+	// exit when raft goes down
+	go func() {
+		if err, ok := <-errorC; ok {
+			log.Fatal(err)
+		}
+		os.Exit(0)
+	}()
 
 	srv := http.Server{
 		Addr: ":" + strconv.Itoa(port),

--- a/contrib/raftexample/raft.go
+++ b/contrib/raftexample/raft.go
@@ -236,8 +236,6 @@ func (rc *raftNode) stop() {
 	close(rc.commitC)
 	close(rc.errorC)
 	rc.node.Stop()
-
-	os.Exit(0)
 }
 
 func (rc *raftNode) stopHTTP() {

--- a/contrib/raftexample/raft.go
+++ b/contrib/raftexample/raft.go
@@ -41,10 +41,11 @@ type raftNode struct {
 	commitC     chan *string             // entries committed to log (k,v)
 	errorC      chan error               // errors from raft session
 
-	id     int      // client ID for raft session
-	peers  []string // raft peer URLs
-	join   bool     // node is joining an existing cluster
-	waldir string   // path to WAL directory
+	id        int      // client ID for raft session
+	peers     []string // raft peer URLs
+	join      bool     // node is joining an existing cluster
+	waldir    string   // path to WAL directory
+	lastIndex uint64   // index of log at start
 
 	// raft backing for the commit/error channel
 	node        raft.Node
@@ -90,8 +91,8 @@ func (rc *raftNode) publishEntries(ents []raftpb.Entry) bool {
 		switch ents[i].Type {
 		case raftpb.EntryNormal:
 			if len(ents[i].Data) == 0 {
-				// ignore conf changes and empty messages
-				continue
+				// ignore empty messages
+				break
 			}
 			s := string(ents[i].Data)
 			select {
@@ -103,7 +104,6 @@ func (rc *raftNode) publishEntries(ents []raftpb.Entry) bool {
 		case raftpb.EntryConfChange:
 			var cc raftpb.ConfChange
 			cc.Unmarshal(ents[i].Data)
-
 			rc.node.ApplyConfChange(cc)
 			switch cc.Type {
 			case raftpb.ConfChangeAddNode:
@@ -116,6 +116,15 @@ func (rc *raftNode) publishEntries(ents []raftpb.Entry) bool {
 					return false
 				}
 				rc.transport.RemovePeer(types.ID(cc.NodeID))
+			}
+		}
+
+		// special nil commit to signal replay has finished
+		if ents[i].Index == rc.lastIndex {
+			select {
+			case rc.commitC <- nil:
+			case <-rc.stopc:
+				return false
 			}
 		}
 	}
@@ -144,19 +153,22 @@ func (rc *raftNode) openWAL() *wal.WAL {
 	return w
 }
 
-// replayWAL replays WAL entries into the raft instance and the commit
-// channel and returns an appendable WAL.
+// replayWAL replays WAL entries into the raft instance.
 func (rc *raftNode) replayWAL() *wal.WAL {
 	w := rc.openWAL()
-	_, _, ents, err := w.ReadAll()
+	_, st, ents, err := w.ReadAll()
 	if err != nil {
 		log.Fatalf("raftexample: failed to read WAL (%v)", err)
 	}
 	// append to storage so raft starts at the right place in log
 	rc.raftStorage.Append(ents)
-	rc.publishEntries(ents)
-	// send nil value so client knows commit channel is current
-	rc.commitC <- nil
+	// send nil once lastIndex is published so client knows commit channel is current
+	if len(ents) > 0 {
+		rc.lastIndex = ents[len(ents)-1].Index
+	} else {
+		rc.commitC <- nil
+	}
+	rc.raftStorage.SetHardState(st)
 	return w
 }
 

